File: backend/app/db/models/forge.py
"""SQLAlchemy models for The Forge — user-built custom bots."""
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import BigInteger, Float, ForeignKey, Text, DateTime
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class UserForgeBot(Base):
    __tablename__ = "user_forge_bots"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(BigInteger, nullable=False, index=True)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(Text, nullable=False, default="active")
    # JSON-encoded lists stored as TEXT (SQLite compatibility)
    strategies: Mapped[str] = mapped_column(Text, nullable=False, default="[]")
    watchlist: Mapped[str] = mapped_column(Text, nullable=False, default="[]")
    capital_pct: Mapped[float] = mapped_column(Float, nullable=False, default=5.0)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    last_scanned_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    last_fired_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    signals: Mapped[list[UserForgeSignal]] = relationship(
        "UserForgeSignal", back_populates="forge_bot", cascade="all, delete-orphan"
    )


class UserForgeSignal(Base):
    __tablename__ = "user_forge_signals"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    forge_bot_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("user_forge_bots.id", ondelete="CASCADE"), nullable=False
    )
    user_id: Mapped[int] = mapped_column(BigInteger, nullable=False, index=True)
    ticker: Mapped[str] = mapped_column(Text, nullable=False)
    strategy_id: Mapped[str] = mapped_column(Text, nullable=False)
    side: Mapped[str] = mapped_column(Text, nullable=False)
    confidence: Mapped[float] = mapped_column(Float, nullable=False)
    entry_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    stop_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    target_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    discord_message_id: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )

    forge_bot: Mapped[UserForgeBot] = relationship("UserForgeBot", back_populates="signals")

File: frontend/src/api/forge.ts
import client from "./client";

export interface ForgeBot {
  id: number;
  name: string;
  description: string | null;
  status: string;
  strategies: string[];
  watchlist: string[];
  capital_pct: number;
  created_at: string;
  last_scanned_at: string | null;
  last_fired_at: string | null;
  signal_count: number;
}

export interface ForgeSignal {
  id: number;
  forge_bot_id: number;
  bot_name: string;
  ticker: string;
  strategy_id: string;
  display_name: string;
  side: string;
  confidence: number;
  entry_price: number | null;
  stop_price: number | null;
  target_price: number | null;
  reason: string | null;
  created_at: string;
}

export interface CreateForgeBotPayload {
  name: string;
  description?: string;
  strategies: string[];
  watchlist: string[];
  capital_pct: number;
}

export const getForgeBots = () =>
  client.get<{ bots: ForgeBot[] }>("/api/forge/bots").then((r) => r.data);

export const createForgeBot = (payload: CreateForgeBotPayload) =>
  client.post<ForgeBot>("/api/forge/bots", payload).then((r) => r.data);

export const updateForgeBot = (id: number, payload: Partial<CreateForgeBotPayload>) =>
  client.put<ForgeBot>(`/api/forge/bots/${id}`, payload).then((r) => r.data);

export const updateForgeBotStatus = (id: number, status: "active" | "paused") =>
  client
    .patch<{ ok: boolean; status: string }>(`/api/forge/bots/${id}/status`, { status })
    .then((r) => r.data);

export const deleteForgeBot = (id: number) =>
  client.delete(`/api/forge/bots/${id}`).then((r) => r.data);

export const getForgeSignals = () =>
  client.get<{ signals: ForgeSignal[] }>("/api/forge/signals").then((r) => r.data);

export const getForgeBotSignals = (id: number) =>
  client.get<{ signals: ForgeSignal[] }>(`/api/forge/bots/${id}/signals`).then((r) => r.data);
